crates/hir_def/src/type_ref.rs

```diff
@@ -290,3 +314,69 @@ impl TypeBound {
         }
     }
 }
+
+pub fn expand_type_ref<'a>(
+    db: &dyn DefDatabase,
+    module_id: ModuleId,
+    type_ref: &'a TypeRef,
+) -> Option<Cow<'a, TypeRef>> {
+    let macro_call = match type_ref {
+        TypeRef::Macro(macro_call) => macro_call,
+        _ => return Some(Cow::Borrowed(type_ref)),
```

Comment:
the only place that calls this function always calls it with a `TypeRef::Macro`, so this should never be hit, but if it were to be hit, that code would recurse infinitely. So how about directly passing the `macro_call` instead, or panicking here? Then this function could also just return an `Option<TypeRef>`. (Actually, we'd want to pass through expansion errors as well, but I think that can be done later.)

Reply:
Done (took the panic approach).